src/cipher/mod.rs

```diff
@@ -0,0 +1,359 @@
+    }
+
+    /// Maximum number of scalars allowed per message
+    pub fn capacity() -> usize {
+        MESSAGE_CAPACITY
+    }
+
+    /// Bytes consumed on serialization of the poseidon cipher
+    pub fn serialized_size() -> usize {
+        ENCRYPTED_DATA_SIZE
+    }
+
+    /// Encrypt a slice of scalars into an internal cipher representation
+    ///
+    /// The message size will be truncated to [`MESSAGE_CAPACITY`] bits
+    pub fn encrypt(
+        message: &[BlsScalar],
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Self {
+        let zero = BlsScalar::zero();
+        let mut strategy = ScalarStrategy::new();
+
+        let mut cipher = [zero; CIPHER_SIZE];
+        let mut state = PoseidonCipher::initial_state(secret, *nonce);
+
+        strategy.perm(&mut state);
+
+        (0..MESSAGE_CAPACITY).for_each(|i| {
+            state[i + 1] += if i < message.len() {
+                message[i]
+            } else {
+                BlsScalar::zero()
+            };
+
+            cipher[i] = state[i + 1];
+        });
+
+        strategy.perm(&mut state);
+        cipher[MESSAGE_CAPACITY] = state[1];
+
+        PoseidonCipher::new(cipher)
+    }
+
+    /// Perform the decrypt of a previously encrypted message.
+    ///
+    /// Will return `None` if the decryption fails.
+    pub fn decrypt(
+        &self,
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Option<[BlsScalar; MESSAGE_CAPACITY]> {
+        let zero = BlsScalar::zero();
+        let mut strategy = ScalarStrategy::new();
+
+        let mut message = [zero; MESSAGE_CAPACITY];
+        let mut state = PoseidonCipher::initial_state(secret, *nonce);
+
+        strategy.perm(&mut state);
+
+        (0..MESSAGE_CAPACITY).for_each(|i| {
+            message[i] = self.cipher[i] - state[i + 1];
+            state[i + 1] = self.cipher[i];
+        });
+
+        strategy.perm(&mut state);
+
+        if self.cipher[MESSAGE_CAPACITY] == state[1] {
+            Some(message)
+        } else {
+            None
+        }
+    }
+
+    /// Getter for the cipher
+    pub fn cipher(&self) -> &[BlsScalar; CIPHER_SIZE] {
+        &self.cipher
+    }
+
+    /// Returns the initial state of the encryption
+    pub fn initial_state(
+        secret: &AffinePoint,
+        nonce: BlsScalar,
+    ) -> [BlsScalar; WIDTH] {
+        [
+            // Domain
+            BlsScalar::from_raw([0x100000000u64, 0, 0, 0]),
+            // Length
+            BlsScalar::from_raw([2u64, 0, 0, 0]),
+            secret.get_x(),
+            secret.get_y(),
+            nonce,
+        ]
+    }
+
+    /// Returns the initial state of the encryption within a composer circuit
+    pub fn initial_state_circuit(
+        composer: &mut StandardComposer,
+        ks0: Variable,
+        ks1: Variable,
+        nonce: Variable,
+    ) -> [Variable; WIDTH] {
+        let domain = BlsScalar::from_raw([0x100000000u64, 0, 0, 0]);
+        let domain = composer.add_input(domain);
```

Comment:
If this is always fixed, it should be added to the circuit description as a selector polynomial coefficient. (Use it as `q_c` / `q_o` / `q_l` / `q_m` or `qr_r` instead of as a `Variable`.

Reply:
This is the initial input of the hades permutation. Such change would imply changes in hades, which is undesirable.

Reply:
KK, then at least constrain it to a constant with the method I mentioned for the `zero` also.